src/mscorlib/src/System/Threading/Tasks/ValueTask.CoreCLR.cs

```diff
@@ -0,0 +1,17 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Threading.Tasks
+{
+    readonly partial struct ValueTask<TResult>
+    {
+        private static void ThrowNullTaskException() => ThrowHelper.ThrowArgumentNullException(ExceptionArgument.task);
+
+        private bool IsTaskCompletedSuccessfully => _task.IsCompletedSuccessfully;
+
+        private Task<TResult> GetTaskForResult() => AsyncTaskMethodBuilder<TResult>.GetTaskForResult(_result);
```

Comment:
Are these extra layers going to regress perf? Note that the JIT is not great at inlining a complex generic constructs like these.

Reply:
I'll check.

Reply:
If it ends up being a problem, I expect the best recourse would be to just ifdef it.  Is there an ifdef we use for this kind of thing? `#if CORECLR`?

Reply:
I think we should be using `#if netstandard` for the less efficient portable version since the portable version is the odd one. Everything else will want to use the more efficient version -  CORECLR, CORERT, MONO.